### PR TITLE
curvefs/client: adapter disk cache size

### LINKED ISSUE #757: curvefs: we want to limit the disk cache size
**Is your feature request related to a problem? (你需要的功能是否与某个问题有关?)**

adaptive disk cache size.

**Describe the solution you'd like (描述你期望的解决方法)**

**Describe alternatives you've considered (描述你想到的折衷方案)**

**Additional context/screenshots (更多上下文/截图)**

### PR DESCRIPTION
**Pull request by wuhongsong**
add the function that control the size of disk cache.

### What problem does this PR solve?

Issue Number: close #757  

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by wuhongsong** [comment]
recheck

File: curvefs/conf/client.conf [hunk 1]
```diff
@@ -137,6 +137,10 @@ diskCache.asyncLoadPeriodMs=5
 # util less than safeRatio
 diskCache.fullRatio=90
 diskCache.safeRatio=70
+# the max size disk cache can use
+diskCache.maxUsableSpaceBytes=107374182400;
```
Comment (xu-chaojie) [inline curvefs/conf/client.conf:141]: do not need ";" here

File: curvefs/conf/client.conf [hunk 2]
```diff
@@ -137,6 +137,10 @@ diskCache.asyncLoadPeriodMs=5
...
+# the max time system command can run
+diskCache.cmdTimeoutSec=300;
```
Comment (xu-chaojie) [inline curvefs/conf/client.conf:143]: do not need ";" here

File: curvefs/src/client/s3/disk_cache_manager.cpp [hunk 3]
```diff
@@ -189,32 +207,43 @@ int64_t DiskCacheManager::CacheDiskUsedRatio() {
...
-    if (ratio >= fullRatio_) {
+    usedBytes_.fetch_add(std::stod(result), std::memory_order_seq_cst);
+    VLOG(3) << "cache disk used size is: " << result;
+    return;
+}
+
+bool DiskCacheManager::IsDiskCacheFull() {
+    int64_t ratio = diskFsUsedRatio_.load(std::memory_order_seq_cst);
+    uint64_t usedBytes = GetDiskUsedbytes();
+    if (ratio >= fullRatio_ ||
```
Comment (baijiaruo) [inline curvefs/src/client/s3/disk_cache_manager.cpp:232]: you already have diskusedbytes, why use ratio?

File: curvefs/src/client/s3/disk_cache_manager.cpp [hunk 4]
```diff
@@ -189,32 +207,43 @@ int64_t DiskCacheManager::CacheDiskUsedRatio() {
...
+    std::string cmd = "timeout " + std::to_string(cmdTimeoutSec_) +
+        " du -sb " + cacheDir_ + " | awk '{printf $1}' ";
+    SysUtils sysUtils;
+    std::string result = sysUtils.RunSysCmd(cmd);
+    if (result.empty()) {
+        LOG(ERROR) << "get disk used size failed.";
+        return;
     }
-    if (ratio >= fullRatio_) {
+    usedBytes_.fetch_add(std::stod(result), std::memory_order_seq_cst);
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/disk_cache_manager.cpp:224]: std::stod will throw an exception when “result” is invalid， use another function in curve/common instead

File: curvefs/src/client/s3/disk_cache_manager.cpp [hunk 5]
```diff
@@ -174,7 +192,7 @@ int DiskCacheManager::LinkWriteToRead(const std::string fileName,
     return cacheRead_->LinkWriteToRead(fileName, fullWriteDir, fullReadDir);
 }
 
-int64_t DiskCacheManager::CacheDiskUsedRatio() {
+int64_t DiskCacheManager::SetDiskFsUsedRatio() {
     struct statfs stat;
     if (posixWrapper_->statfs(cacheDir_.c_str(), &stat) == -1) {
```
[the remark this answers is not in the document]
Comment (wuhongsong) [reply curvefs/src/client/s3/disk_cache_manager.cpp:197]: of course can not

File: curvefs/src/common/utils.h [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+#include <string>
+
+namespace curvefs {
+namespace common {
+
+class SysUtils {
+ public:
+    SysUtils() {}
+    virtual ~SysUtils() {}
```
Comment (xu-chaojie) [inline curvefs/src/common/utils.h:35]: virtual is not needed

File: curvefs/src/client/s3/disk_cache_manager.cpp (hunk 3 above)
Comment (wuhongsong) [reply curvefs/src/client/s3/disk_cache_manager.cpp:232]: when we use the disk that not separate，especially the system disk, we prefer not to write when the disk is near full

File: curvefs/src/client/s3/disk_cache_manager.cpp [hunk 7]
```diff
@@ -277,24 +307,33 @@ void DiskCacheManager::TrimCache() {
...
                 ret = posixWrapper_->remove(toDelFile);
                 if (ret < 0) {
                     LOG(ERROR)
-                        << "remove disk file error, file = " << *cacheKeyIter;
+                        << "remove disk file error, file is: " << *cacheKeyIter;
                     cachedObjNameTmp.erase(cacheKeyIter);
                     continue;
                 }
-                VLOG(3) << "remove disk file success, file = " << *cacheKeyIter;
+                DecDiskUsedBytes(statReadFile.st_size);
```
Comment (xu-chaojie) [inline curvefs/src/client/s3/disk_cache_manager.cpp:334]: Is it possible that statReadFile has holes?   SetDiskInitUsedBytes use “du”  to calc disk used， which is not conclude the size of hole？

Comment (wuhongsong) [reply curvefs/src/client/s3/disk_cache_manager.cpp:339]: the file has no hole

**Review by baijiaruo — APPROVED** [review]

**Review by xu-chaojie — APPROVED** [review]

**Pull request merged**